plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-product-related-products-controller.php

```diff
@@ -0,0 +1,254 @@
+			),
+			'default'           => array(),
+			'validate_callback' => 'rest_validate_request_arg',
+			'sanitize_callback' => 'wp_parse_id_list',
+		);
+
+		// Add a parameter to whether combine or not the categories and tags.
+		$params['combine'] = array(
+			'description'       => __( 'Combine the product terms (categories, tags, ...).', 'woocommerce' ),
+			'type'              => 'boolean',
+			'default'           => false,
+			'validate_callback' => 'rest_validate_request_arg',
+			'sanitize_callback' => 'rest_sanitize_boolean',
+		);
+
+		return $params;
+	}
+
+	/**
+	 * Get the related product categories.
+	 *
+	 * @param array $cats The related product categories.
+	 * @return array
+	 */
+	public function get_related_product_cat_terms( $cats ) {
+		if ( ! $this->combine ) {
+			return $this->categories;
+		}
+
+		/*
+		 * Merge the given terms with the related product categories
+		 * ensuring to remove any duplicates.
+		 */
+		return array_unique( array_merge( $cats, $this->categories ) );
+	}
+
+	/**
+	 * Get the related product tags.
+	 *
+	 * @param array $tags The related product tags.
+	 * @return array
+	 */
+	public function get_related_product_tag_terms( $tags ) {
+		if ( ! $this->combine ) {
+			return $this->tags;
+		}
+
+		return array_unique( array_merge( $tags, $this->tags ) );
+	}
+
+	/**
+	 * Prepare objects query.
+	 *
+	 * @since  3.0.0
+	 * @param  WP_REST_Request $request Full details about the request.
+	 * @return array
+	 */
+	protected function prepare_objects_query( $request ) {
+		$args = parent::prepare_objects_query( $request );
+
+		$args['post__in'] = $this->ids;
+		return $args;
+	}
+
+	/**
+	 * Get the related products.
+	 * Related products is handled by the wc_get_related_products() core function.
+	 * Currently, the data that define the related products are the categories and tags.
+	 * This endpoint accepts and combines the categories and tags parameters.
+	 *
+	 * @param WP_REST_Request $request Request object.
+	 * @return object
+	 */
+	public function get_items( $request ) {
+		$id     = $request->get_param( 'id' );
+
+		$categories = $request->get_param( 'categories' );
+		$tags       = $request->get_param( 'tags' );
+		$combine    = $request->get_param( 'combine' );
+
+		if ( ! empty( $combine ) ) {
+			$this->combine = $combine;
+		}
+
+		/*
+		 * If categories param is defined,
+		 * filter the related products by the categories.
+		 */
+		if ( ! empty( $categories ) ) {
+			$this->categories = $categories;
+			add_filter(
+				'woocommerce_get_related_product_cat_terms',
+				array(
+					$this,
+					'get_related_product_cat_terms',
+				)
+			);
+		}
+
+		/*
+		 * If tags param is defined,
+		 * filter the related products by the tags.
+		 */
+		if ( ! empty( $tags ) ) {
+			$this->tags = $tags;
+			add_filter(
+				'woocommerce_get_related_product_tag_terms',
+				array(
+					$this,
+					'get_related_product_tag_terms',
+				)
+			);
+		}
+
+		$this->ids = wc_get_related_products( $id );
```

Comment:
It sucks that this function is totally dependent on a product id. Otherwise I would of suggested us having two routes:
`products/(?P<id>[\d]+)/related-products` - for getting the data specific to the product
`products/related-products` - for getting the data based on the passed in tags and categories.

Reply:
Having both endpoints sounds good to me.
The core function isn't a big deal IMO, We can hack a ghost product 👻 or create a new one.
Also, I'd rename the generic one since it isn't returning relayed [products but suggesting ones. So...

```
products/suggested-products
````

will return suggestions based on categories, tags, ....

Reply:
Nice, yeah `suggested` makes indeed more sense.